### PR TITLE
Refactor and remove `SamplerBase` class

### LINKED ISSUE #2607: ``BaseSampler`` vs ``SamplerBase`` classes
There are both ``BaseSampler`` and ``SamplerBase`` classes. This issue tracks determining the difference between the two and if we can combine / use just one base class.

### PR DESCRIPTION
**Pull request by angela97lin**
Closes #2607 

Removes `SamplerBase` class, in favor of `BaseSampler`. `BalancedClassificationSampler` was a subclass of `SamplerBase` and only used as a component object of the Undersampler class, so simplified by removing the `BalancedClassificationSampler` class and moving the implementation to `Undersampler`.

### CODE REVIEW

File: evalml/pipelines/components/transformers/samplers/undersampler.py [hunk 1]
```diff
@@ -42,12 +41,32 @@ def __init__(
         random_seed=0,
         **kwargs,
     ):
+        if sampling_ratio <= 0 or sampling_ratio > 1:
```
Comment (chukarsten) [inline evalml/pipelines/components/transformers/samplers/undersampler.py:44]: Is it worth getting ahead of the game and adding "raises" for these newish exceptions to the docstring?

File: evalml/pipelines/components/transformers/samplers/undersampler.py [hunk 2]
```diff
@@ -59,20 +78,98 @@ def _initialize_sampler(self, X, y):
...
+        indices = self.fit_resample(X_ww, y_ww)
 
         train_indices = index_df[index_df.isin(indices)].index.values.tolist()
         return X_ww.iloc[train_indices], y_ww.iloc[train_indices]
+
+    def _find_ideal_samples(self, y):
+        """Returns dictionary of examples to drop for each class if we need to resample.
+
+        Arguments:
+            y (pd.Series): Target data passed in
```
Comment (chukarsten) [inline evalml/pipelines/components/transformers/samplers/undersampler.py:95]: hypernit - "."

File: evalml/pipelines/components/transformers/samplers/undersampler.py [hunk 3]
```diff
@@ -59,20 +78,98 @@ def _initialize_sampler(self, X, y):
...
+        if len(self.sampling_ratio_dict):
+            result = self._sampling_dict_to_remove_dict(y)
+        else:
+            result = self._find_ideal_samples(y)
+        indices_to_drop = []
+        if len(result):
+            # iterate through the classes we need to undersample and remove the number of samples we need to remove
+            for key, value in result.items():
+                indices = y.index[y == key].values
+                # import pdb; pdb.set_trace()
```
Comment (chukarsten) [inline evalml/pipelines/components/transformers/samplers/undersampler.py:168]: Can we get rid of this while we're here?  I would blame this on me.

**Review by chukarsten — APPROVED** [review]
Just a few hypernits and an ask for an unnecessary comment cleanup!  But awesome, this looks good.

File: evalml/pipelines/components/transformers/samplers/undersampler.py (hunk 3 above)
Comment (angela97lin) [reply evalml/pipelines/components/transformers/samplers/undersampler.py:168]: LOL oops

File: evalml/pipelines/components/transformers/samplers/undersampler.py (hunk 1 above)
Comment (angela97lin) [reply evalml/pipelines/components/transformers/samplers/undersampler.py:44]: Love it! Unfortunately, the linting packages aren't smart enough to tell that the class docstring should be connected to the __init__ docstring but going to add this 😁

**Pull request merged**